ui/radar/radar.ts

```diff
@@ -0,0 +1,499 @@
+    this.options = Options;
+    this.hunts = Object.assign({}, HuntData, Options.CustomMonsters);
+    this.lang = this.options.ParserLanguage ?? 'en';
+    this.nameToHuntEntry = {};
+    this.regexes = {
+      abilityFull: NetRegexes.abilityFull(),
+      addedCombatantFull: NetRegexes.addedCombatantFull(),
+      instanceChanged: instanceChangedRegexes[this.options.ParserLanguage] || instanceChangedRegexes['en'],
+      wasDefeated: NetRegexes.wasDefeated(),
+    };
+
+    for (const hunt of Object.values(this.hunts)) {
+      let name = hunt.name;
+      if (typeof name === 'object' && !Array.isArray(name))
+        name = name[this.lang] ?? name['en'];
+
+      if (typeof name === 'string') {
+        this.nameToHuntEntry[name.toLowerCase()] = hunt;
+      } else {
+        for (const nameItem of name)
+          this.nameToHuntEntry[nameItem.toLowerCase()] = hunt;
+      }
+    }
+  }
+
+  AddMonster(log: string, hunt: HuntEntry,
+      matches: Matches<NetRegex<AddedCombatantFullParams>>) {
+    if (!this.playerPos)
+      return;
+    if (!matches)
+      return;
+    if (matches.id === undefined ||
+        matches.name === undefined ||
+        matches.npcNameId === undefined ||
+        matches.hp === undefined ||
+        matches.currentHp === undefined ||
+        matches.x === undefined ||
+        matches.y === undefined ||
+        matches.z === undefined)
+      throw new UnreachableCode();
+
+
+    if (hunt.id && matches.npcNameId !== hunt.id)
+      return;
+    if (hunt.regex && !hunt.regex.test(log))
+      return;
+    if (hunt.hp && parseFloat(matches.hp) < hunt.hp)
+      return;
+    if (matches.currentHp === '0') // hunt is already dead
+      return;
+
+    let options = this.options;
+    // option overwrite
+    if (hunt.rank && hunt.rank in options.RankOptions)
+      options = Object.assign({}, this.options, options.RankOptions[hunt.rank]);
+    // Check if monster rank enabled or not in config
+    if (hunt.rank === 'B' && !options.BRankEnabled)
+      return;
+    if (options.Enabled === false)
+      return;
+    if (options.Type === 'mob') {
+      if (!matches.id.startsWith('4'))
+        return;
+      if (!matches.npcNameId || matches.npcNameId === '0')
+        return;
+    }
+
+    const mobKey = matches.name.toLowerCase();
+    if (!mobKey)
+      return;
+    const targetMob = this.targetMonsters[mobKey];
+    if (targetMob) {
+      // Get positions
+      const playerPos = new Point2D(this.playerPos.x, this.playerPos.y);
+      const oldPos = targetMob.pos;
+      const newPos =
+        new Point2D(parseFloat(matches.x), parseFloat(matches.y));
+
+      // Calculate distances
+      const oldDistance = playerPos.distance(oldPos);
+      const newDistance = playerPos.distance(newPos);
+
+      // Update position only if its closer than the current one
+      if (newDistance < oldDistance) {
+        targetMob.pos = newPos;
+        targetMob.posZ = parseFloat(matches?.z ?? '0');
```

Comment:
```suggestion
        targetMob.posZ = parseFloat(matches.z ?? '0');
```
I don't think matches can be undefined?